src/3d/qgscameracontroller.cpp

```diff
@@ -307,6 +307,28 @@ void QgsCameraController::onPositionChanged( Qt3DInput::QMouseEvent *mouse )
   }
 }
 
+void QgsCameraController::screenPointToWorldPos( QPoint position, Qt3DRender::QCamera *cameraBefore,
+    std::function<void( double, const QVector3D & )> onSuccess )
+{
+  double depth = sampleDepthBuffer( mDepthBufferImage, position.x(), position.y() );
+  if ( !std::isfinite( depth ) )
+  {
+    QgsDebugMsgLevel( QStringLiteral( "screenPointToWorldPos: depth is NaN or Inf. This should not happen." ), 2 );
+  }
+  else
+  {
+    QVector3D worldPosition = Qgs3DUtils::screenPointToWorldPos( position, depth, mViewport.size(), cameraBefore );
+    if ( !std::isfinite( worldPosition.x() ) || !std::isfinite( worldPosition.y() ) || !std::isfinite( worldPosition.z() ) )
+    {
+      QgsDebugMsgLevel( QStringLiteral( "screenPointToWorldPos: position is NaN or Inf. This should not happen." ), 2 );
+    }
+    else
+    {
+      onSuccess( depth, worldPosition );
+    }
+  }
+}
```

Comment:
To me this could be simplified without the use of `onSuccess` callback function. Why not just have true/false returned from the function, and then the client code can do 
```cpp
if ( screenPointToWorldPos( ... ) )
{
   // success - do something
}
```
That would be easier to read than having to deal with some lambda functions, and it will be also possible to respond to failure (with the current approach there's no callback for failure)

Reply:
The main idea with the lambda function was to avoid declare variables outside of `screenPointToWorldPos` which might never been used (`depth` and `worldPosition`): 

```c++
double depth;
QVector3D worlPosition;
if ( screenPointToWorldPos( mMousePos, mCameraBeforeZoom.get(), depth, worldPosition ) )
{
 // success - do something
}
```

Reply:
I ended up using reference. Using a pointer for `depth` added more complexity to `screenPointToWorldPos`, it was not worth it.